cirq-core/cirq/transformers/target_gatesets/compilation_target_gateset.py

```diff
@@ -115,3 +115,95 @@ def postprocess_transformers(self) -> List['cirq.TRANSFORMER']:
             drop_negligible.drop_negligible_operations,
             drop_empty_moments.drop_empty_moments,
         ]
+
+
+class TwoQubitCompilationTargetGateset(CompilationTargetGateset):
+    """Abstract base class to create two-qubit target gatesets.
+
+    This base class can be used to create two-qubit compilation target gatesets. It automatically
+    implements the logic to
+
+        1. Expand composite gates acting on > 2 qubits using
+        2. Merge connected components of 1 and 2 qubit unitaries into tagged circuit operations.
+        3. Use `_decompose_single_qubit_operation` and `_decompose_two_qubit_operation` to figure
+        out how to rewrite the merged connected components using gates from this gateset.
+        4. Replace connected components with inefficient implementations (higher number of 2q
+        interactions) with more efficient rewrites to minimize total number of 2q interactions.
+        5. Apply cleanup transformations to preserve moment structure (eg: alternating layers of
+        single/two qubit gates) and minimize circuit depth (eg: drop negligible operations and
+        empty moments)
+
+    Derived classes should simply implement the `_decompose_two_qubit_operation` method and provide
+    analytical decomposition of any 2q unitary using gates from the target gateset.
+    """
+
+    @property
+    def num_qubits(self) -> int:
+        return 2
+
+    def decompose_to_target_gateset(self, op: 'cirq.Operation', moment_idx: int) -> DecomposeResult:
+        if not 1 <= protocols.num_qubits(op) <= 2:
+            return self._decompose_multi_qubit_operation(op, moment_idx)
+        if protocols.num_qubits(op) == 1:
+            return self._decompose_single_qubit_operation(op, moment_idx)
+        op_untagged = op.untagged
+        old_optree = (
+            [*op_untagged.circuit]
+            if isinstance(op_untagged, circuits.CircuitOperation)
+            and self._intermediate_result_tag in op.tags
+            else [op]
+        )
+        old_2q_gate_count = len([o for o in ops.flatten_to_ops(old_optree) if len(o.qubits) == 2])
```

Comment:
Nit: if flatten_ops produces a generator, we can just do a sum instead of needing to make a temp list:

```python3
old_2q_gate_count = sum(1 for o in ops.flatten_to_ops(...) if len(op.qubits) == 2)
```

Reply:
done